### DISCUSSION TITLE
Load-bearing comments

### DISCUSSION THREAD

**Discussion by passcod** [kdl-org/kdl#68] [Ideas]
This is perhaps more aimed at implementations than the spec, but I think it would be really beneficial to have first-class support for comments **in the APIs** rather than them being silently discarded.

The main use would be when generating kdl strings from a program:
 - to be able to preserve or insert comments when transforming a document
 - to be able to include comments when making a new one (e.g. writing a default configuration to disk on first run)
 - autoformatter, to an extent: round-trip kdl through a parser and generator and get the "canonical" form out (if that's specified at some point)

This raises questions:
1. slash dash comments are straightforward, as you already need to parse their contents to know where they stop. Output could be a node/attribute/value struct or object with the property commented=true, or something like that. However, representing the other comment types may need a separate object? It doesn't make sense to have a slash slash comment API object have children and attributes and values, surely.
2. How to represent nested comments?
3. Should non slash dash comments be parsed with an indicator of their format (slash slash or slash star)?

Regarding comments in general, there's currently no concept of attaching comments to nodes, comments are just free. E.g. semantically or structurally, should there be a difference between these two below:

```kdl
general {
    colour always
}

// todo: figure out keybindings

plugins discovery=true auto-enable=false {
    inotify enable=true

    // linux: more accurate but requires permissions
    fanotify enable=on-demand
}
````

**Comment by hclarke** [comment]
for tools (formatters, templating, etc.), you're going to need more than just comments in order to do a round trip:
- strings can be raw (with number of `#`), or normal with escape sequences
- identifiers can be bare or strings (with same string things above)
- numbers can be binary/octal/hex, have underscores, leading +/-, etc.
- whitespace also comes in many shapes/sizes
- separators can be newlines or semicolons

And, the common case would be programs using KDL as a config file and such, which don't need to know about comments (and maybe shouldn't, to discourage putting data in comments?)

but also all of this is kinda necessary for tooling. i'm not sure if that means a second parser, or feature flags, or a function to convert to the simpler format. probably depends on the implementation language?

**Reply by passcod** [reply]
All very good points!

Another related question then is whether the spec should define some kind of standard for formatting a kdl document...

**Reply by hkolbeck-streem** [reply]
IMO an exact roundtrip is almost never required, because the use case is almost always creation of some canonical form. At least for quotes it seems like the canonical version would be the least-quoted one, which is what I did [in the Java parser](https://github.com/hkolbeck/kdl4j/blob/trunk/src/main/java/dev/hbeck/kdl/objects/PrintUtil.java#L16).

**Comment by zkat** [comment]
This is definitely an implementation concern, imo. For example, a serde-based parser might not want to deal with comments at all, since it only cares about data coming in, and would have some default way of dealing with data going out.

What you're talking about is what I've heard called a "document-oriented parser", which _would_ retain comment, spacing, and input format.

I don't know if there's anything to do to allow this at the specification level, although I already made some decisions that require this type of parser for proper round-tripping for sure. But unless you're something as simple as JSON, this will usually be the case.

**Comment by jamesmunns** [comment]
Just noting that I found this while trying to see if there was a way to view comments on a parsed `KdlDocument`/`KdlNode`. I'm using kdl as the base format of an [IDL](https://github.com/jamesmunns/postcard-idl/blob/main/input/input-001.kdl), and it would be neat to be able to annotate types/etc with semantically attached comments.

Would you have interest in a PR that added support for Rust-style doc comments?

Something like:

```kdl
types {
    //! This is a module doc comment, and would be attached to the `types` node!

    // this is a normal comment, not attached to any parsed nodes

    /// This is doc comment, and would be attached to the `struct` node!
    struct "Rgb8" {
        /// This is a doc comment, and would be attached to the `r` node!
        r "u8"
        /// This is green
        g "u8"
        /// This is blue
        b "u8"
    }
}
```

If there's interest, I can look into whether this would be feasible to support? From a standard perspective, I *think* it wouldn't interfere with anything, since there's no mandatory space character, so these would "just" be fancy line comments according to the spec.

**Reply by zkat** [reply]
This is already supported by virtue of them being standard line comments. I think it’s ok to have your IDL post processors
 look at the content of the comments and act accordingly.

**Reply by jamesmunns** [reply]
Ah sorry, I missed that this was on the `kdl` repo discussions, and not `kdl-rs`, which is what my IDL parser is using. Would you like me to open an issue over there to discuss that as a feature?

**Reply by zkat** [reply]
I would say the same thing: kdl-rs doesn’t do any high level processing of comments right now. It treats them as whitespace. But you should be able to just throw another parsing layer at it as part of processing your IDL

additionally: in your example, none of those strings need to be quoted in v2. I think that would make your DSL look nicer:

```kdl

types {
    //! This is a module doc comment, and would be attached to the `types` node!

    // this is a normal comment, not attached to any parsed nodes

    /// This is doc comment, and would be attached to the `struct` node!
    struct Rgb8 {
        /// This is a doc comment, and would be attached to the `r` node!
        r u8
        /// This is green
        g u8
        /// This is blue
        b u8
    }
}
```

**Reply by jamesmunns** [reply]
Yep, that was my ask: would you be open to grabbing the comments while parsing, and attaching them to `KdlNode`s so they could be grabbed in a single pass of parsing?

re: the quoting, I definitely agree for builtins, some of the types use `<>` pairs, and tuples use `()` pairs, so I went with type quoting for consistency. I also based a lot of my earlier prototyping off of Gankra's `kdl-script`, which also uses string quoting, though I haven't seen if all of my examples work in that case.

My full current test input (as well as the compiler itself) is here if you want to peek: https://github.com/jamesmunns/postcard-idl/blob/main/input/input-001.kdl

**Reply by zkat** [reply]
The comments [are already grabbed by `kdl-rs`](https://docs.rs/kdl/latest/kdl/struct.KdlNodeFormat.html#structfield.leading). You can do your own post-processing on them. I would rather not do further parsing on these, because that goes into DSL-specific whitespace semantics--and this is very much DSL-specific.

**Reply by jamesmunns** [reply]
I totally missed that the comments were parsed into the Format data. Thanks a ton! I'll take a look and see how I missed that.
